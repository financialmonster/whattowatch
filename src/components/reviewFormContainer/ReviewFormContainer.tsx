import React, { FC, useCallback, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import { useFetchReview } from 'hooks/useFetchReview';
import { ReviewForm } from 'components/reviewForm/ReviewForm';
import { TUserReview } from 'types';
import { Notification } from 'components/notification/Notification';
import { reviewsActions } from 'domains/reviews/reviewsActions';

export const ReviewFormContainer: FC = () => {
    const {id} = useParams();
    const dispatch = useDispatch();
    const {isReviewFetching, reviewError, reviewFormSubmitHandler} = useFetchReview();
    const verifiedIsReviewFetching = (isReviewFetching as boolean);

    useEffect(() => {
        dispatch(reviewsActions.resetReviewError());
    }, [dispatch]);

    const handleReviewFormSubmit = useCallback((userReview: TUserReview) => {
        reviewFormSubmitHandler(+id, userReview);
    }, [id, reviewFormSubmitHandler]);

    if(reviewError) {
        return <div>{(reviewError as Error).message}</div>;
    }

    return (
        <>
            <ReviewForm isReviewFetching={verifiedIsReviewFetching} reviewFormSubmitHandler={handleReviewFormSubmit} />
            {(isReviewFetching) &&
                <Notification>
                    Sending your review... Please, wait
                </Notification>
            }
        </>
    );
}
