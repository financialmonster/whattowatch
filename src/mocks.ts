import { Map, fromJS, List } from 'immutable';

export const createStoreMock = (mode: string = `common`) => {
    return {
        promo: Map({
            promo: Map(),
            isPromoFetching: true,
            promoError: null
        }),
        films: Map({
            films: fromJS([{genre: `Thriller`, id: 1}, {genre: `Comedy`, id: 2}]),
            isFilmsFetching: mode === `detailed`,
            filmsError: null,
            filter: `All genres`
        }),
        auth: Map({
            user: null,
            authError: null,
            isAuthFetching: false,
            authStatusError: null,
            isAuthStatusFetching: false
        }),
        reviews: Map({
            reviews: List(),
            isReviewsFetching: true,
            reviewsError: null
        })
    }
}
