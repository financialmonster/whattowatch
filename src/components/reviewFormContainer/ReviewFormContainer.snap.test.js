/* eslint-disable import/first */
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({id: 15})}
));

import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { ReviewFormContainer } from './ReviewFormContainer';
import { createStoreMock } from 'mocks';

it(`ReviewFormContainer: correctly rendered`, () => {
    const mockStore = configureStore();
    const store = mockStore(createStoreMock());

    const tree = renderer.create(
        <Provider store={store}>
            <ReviewFormContainer />
        </Provider>        
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
