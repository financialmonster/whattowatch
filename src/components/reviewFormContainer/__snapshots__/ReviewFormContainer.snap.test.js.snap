// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewFormContainer: correctly rendered 1`] = `
<form
  action="#"
  className="add-review__form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
    >
      <input
        checked={false}
        className="rating__input"
        id="star-5"
        name="rating"
        onBlur={[Function]}
        onChange={[Function]}
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating-5
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-4"
        name="rating"
        onBlur={[Function]}
        onChange={[Function]}
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating-4
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-3"
        name="rating"
        onBlur={[Function]}
        onChange={[Function]}
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating-3
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-2"
        name="rating"
        onBlur={[Function]}
        onChange={[Function]}
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating-2
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-1"
        name="rating"
        onBlur={[Function]}
        onChange={[Function]}
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating-1
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      name="review-text"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Review text"
      value=""
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        title="Post"
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
</form>
`;
